role-based boot modes: add content

3mkusiak opened the pull request:
Add content for "role-based boot modes" chapter.

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -66,6 +66,35 @@
+* Legacy boot mode - allows for booting alternate OS.
+
+#### References
+[^1]: [user-sysmaint-split](https://www.kicksecure.com/w/index.php?title=Dev/user-sysmaint-split)
+[^2]: [ChromeOS-Boot-Modes](https://docs.mrchromebox.tech/docs/boot-modes/)
```
```suggestion
[^1]: [user-sysmaint-split](https://www.kicksecure.com/w/index.php?title=Dev/user-sysmaint-split)
[^2]: [ChromeOS Boot Modes](https://docs.mrchromebox.tech/docs/boot-modes/)
```

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -66,6 +66,35 @@
+role-base boot, is ChromeOs. ChromeOs includes following boot modes[^2]:
+* Normal mode - default boot option, performs full verification of firmware and
+OS components.
+* Recovery mode - Used for OS and RW firmware components repair.
```
I'm not sure what's `RW` in this case? read-write? If so then you should use lower case and probably whole words. If it's used in more places then first use could be `rw (read-write)` and you could skip `(read-write)` after that.

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -66,6 +66,35 @@
 
+Role-Based boot is a term proposed by Kicksecure, it is a concept of booting
+operating system into different modes based on user roles in order to enhance
+security by isolating executable activities and permissions. The aim for
```
> isolating executable activities and permissions

* isolating permissions - does this mean file permissions are different based on role? Or are you booting into different account with different permissions. Or maybe same account but different group.
* isolating executable activities - that should probably be explained more or rewritten as I'm not sure what it means. What kind of isolation? Sandbox? Maybe you meant process isolation? But isolated from what?

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -66,6 +66,35 @@
+Role-based boot is implemented in Kicksecure and Whonix operating systems.
+Currently supported boot modes are[^1]:
+* `PERSISTENT mode` - intended for performing daily activities (browsing,
+email, chat) with write access only for `/home` directory.
```
Haven't used Kicksecure but what about e.g. `/tmp`?

ArrayBolt3 replied:
I think it's a bit confusing to use write access to distinguish between the various modes - the terms "PERSISTENT mode USER", "LIVE mode USER", and "PERSISTENT mode SYSMAINT" have two different concepts involved in each "mode" - the persistence of changes and the user account in use. "PERSISTENT mode USER" provides full file persistence, and provides normal user access from a permissions standpoint. Normal users in this context are prevented from escalating to root by denying the ability to even execute privilege escalation tools including but not limited to `sudo` and `pkexec`. "LIVE mode USER" is the same as "PERSISTENT mode USER" but the root filesystem (and by extension the /home directory, if it is located on the same filesystem as the root directory) is non-persistent and changes are lost on reboot. "PERSISTENT mode SYSMAINT" provides full persistence, but grants sysmaint access from a permissions standpoint. In practice, this means that a special user named "sysmaint" has its password unlocked, so that it can be signed into. The end-user is discouraged from signing into other user accounts at this time. The "sysmaint" user account is allowed to run privilege escalation tools, thus they can run system maintenance tasks like software updates.

Anything that a normal user would usually be able to write to without having to elevate to root or switch users will still be writable in either "PERSISTENT mode USER" or "LIVE mode USER", just like on any other traditional (mutable) Linux system. This includes the user's temp directory, and also includes directories shared among multiple users, like `/tmp` and `/dev/shm`.

3mkusiak replied:
I elaborated on this based on your feedback. Let me know if it's sufficient.

m-iwanicki reviewed (changes requested):
I think it would be good addition to add how role-based modes is implemented.
The only information I got from this section is very short description of different boot modes in Kicksecure (which https://www.kicksecure.com/wiki/Dev/user-sysmaint-split#Boot_Modes_Comparison_Table shows better but I'm not sure it's correct (it shows that `/home` is read only which conflicts with https://www.kicksecure.com/wiki/Live_Mode#Functionality_Test_of_Live_mode).

Is this supposed to replace <https://www.kicksecure.com/wiki>? If yes then maybe some information from <https://www.kicksecure.com/wiki/Live_Mode> e.g. how to boot into different modes would be good? Unless this section should just have explanation of role-based boot modes instead of how to use it with Kicksecure?

3mkusiak commented:
> The only information I got from this section is very short description of different boot modes in Kicksecure.

Correct, that's why I started this chapter the way I did. The "role-based" in "role-based boot modes" is not (yet) industry standard term nor concept.

> Is this supposed to replace https://www.kicksecure.com/wiki?

I don't know the answer.  I assumed [user-sysmain-split](https://www.kicksecure.com/wiki/Dev/user-sysmaint-split#Boot_Modes_Comparison_Table) would go in detail about how it works, and here we would present only the concept.

macpijan commented on verified_boot_main.md:
```diff
@@ -66,6 +66,35 @@
 
 ### Role-based boot modes
 
+Role-Based boot is a term proposed by Kicksecure, it is a concept of booting
```
I guess we should add https://www.kicksecure.com/wiki/Dev/user-sysmaint-split as an introduction of this concept?

How mature is this feature @adrelanos , or this is mostly a concept / in development right now?

adrelanos replied:
Pretty mature. It's currently being tested and will be included in the next release. Most likely become the default once we rebased Kicksecure on Debian trixie (once blessed stable).

macpijan commented:
> Is this supposed to replace https://www.kicksecure.com/wiki?

Definitely not all of it.

If there is an external reference in the Kicksecure wiki (besides the https://www.kicksecure.com/wiki/Verified_Boot page), we should link it here, and not repeat too much.

If something in the linked page is not clear, or we think it needs adjustment, we may move it to the issue in this repo, discuss it there, and contribute to the wiki directly. This way, we can reference to the Kicksecure wiki in our document, and clarify some concepts at the source at the same time.

If we create an issue and ask @adrelanos specific questions on the "role-based" mode, I am sure we can get some answers.

adrelanos commented:
More documentation here:

3mkusiak commented:
I updated the introduction to be more clear. 

> I guess we should add https://www.kicksecure.com/wiki/Dev/user-sysmaint-split as an introduction of this concept?

Imo I added the bare minimum here in order to avoid repeating content. I intended this "article" to be about the concept, as [user-sysmaint-split ](https://www.kicksecure.com/wiki/Dev/user-sysmaint-split) is about implementation.

Having said that, I don't know if my idea is right. I'd appreciate detailed info on what should be included here.
> we should add https://www.kicksecure.com/wiki/Dev/user-sysmaint-split as an introduction of this concept?

...is too vague.

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -66,6 +66,41 @@
+Currently supported boot modes are[^1]:
+* `PERSISTENT mode USER` - provides persistency for user files. The
+user operates with standard permissions, but they're not allowed to escalate
+privileges (eg. by executing `sudo` command). All locations remain writeable
```
```suggestion
privileges (e.g. by executing `sudo` command). All locations remain writable
```

3mkusiak replied:
Fixed.

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -66,6 +66,41 @@
+privileges (eg. by executing `sudo` command). All locations remain writeable
+including `/tmp` and `/dev/shm`.
+* `LIVE mode USER` - serves similar tole to `PERSISTENT mode USER` with the
+difference that all not-persistent changes are lost during reboot.
```
> all not-persistent changes are lost during reboot

This applies to all modes (that's what non-persistence means). Maybe reword it to mention that filesystem changes are non-persistent i.e. lost during reboot.

3mkusiak replied:
I attempted to re-sentence [this](https://github.com/3mdeb/verified-boot/pull/10/files#diff-3f3cfef590743af79e56bf6aa2508d5d3ce7f08fcf1a856ff8ecd10d4df3e655R83). Let me know if that's what you had in mind.
